tests/opera/parser/test_tosca.py

```diff
@@ -101,3 +101,41 @@ def test_load_from_csar_subfolder(self, tmp_path, yaml_text):
 
         doc = tosca.load(tmp_path, name)
         assert doc.data_types["my_type"]
+
+    def test_duplicate_import(self, tmp_path, yaml_text):
+        name = pathlib.PurePath("template.yaml")
+        (tmp_path / name).write_text(yaml_text(
+            """
+            tosca_definitions_version: tosca_simple_yaml_1_3
+            imports: [ template.yaml ]
+            """
+        ))
+        tosca.load(tmp_path, name)
+
+    def test_duplicate_imports(self, yaml_ast, yaml_text, tmp_path):
+        name = pathlib.PurePath("sub/folder/file.yaml")
+        (tmp_path / name).parent.mkdir(parents=True)
+        (tmp_path / name).write_text(yaml_text(
+            """
+            tosca_definitions_version: tosca_simple_yaml_1_3
+            imports:
+              - imp1.yaml
+            """
+        ))
+        (tmp_path / "sub/folder/imp1.yaml").write_text(yaml_text(
+            """
+            tosca_definitions_version: tosca_simple_yaml_1_3
+            imports:
+              - imp2.yaml
+            """
+        ))
+        (tmp_path / "sub/folder/imp2.yaml").write_text(yaml_text(
+            """
+            tosca_definitions_version: tosca_simple_yaml_1_3
+            data_types:
+              my_type:
+                derived_from: tosca.datatypes.xml
+            """
+        ))
+
+        tosca.load(tmp_path, name)
```

Comment:
This code does not test anything new because all imports are still at the same level. Besides, no file is imported twice or more times here.

I would suggest a test similar to the one below:

```
def test_imports_from_multiple_levels(self, yaml_text, tmp_path):
    name = pathlib.PurePath("template.yaml")
    (tmp_path / name).write_text(yaml_text(
        """
        tosca_definitions_version: tosca_simple_yaml_1_3
        imports:
          - subfolder/a.yaml
          - subfolder/b.yaml
        """
    ))
    (tmp_path / "subfolder").mkdir()
    (tmp_path / "subfolder/a.yaml").write_text(yaml_text(
        """
        tosca_definitions_version: tosca_simple_yaml_1_3
        imports:
          - b.yaml
        """
    ))
    (tmp_path / "subfolder/b.yaml").write_text(yaml_text(
        """
        tosca_definitions_version: tosca_simple_yaml_1_3
        data_types:
          my_type:
            derived_from: tosca.datatypes.xml
        """
    ))

    tosca.load(tmp_path, name)
```

Reply:
I have included the proposed test. But now as expected the CircleCI pytests fail because the new problem emerged. It says `opera.error.ParseError: [MapWrapper] Duplicate keys 'my_type' found in subfolder/b.yaml:4:3 and subfolder/b.yaml:4:3` which was originally an exception that supported the idea about preventing duplicate keys within imports.

Reply:
This should work, which means that our "do not import things twice" implementation is bad. The general idea here is that loader works this way:

 1. Load `template.yaml` and add it to the set of processed files.
 2. Load `subfolder/a.yaml` and add it to the set of processed files.
 3. Load `subfolder/b.yaml` (from `subfolder/a.yaml`) and add it to the set of processed files.
 4. Skip loading `subfolder/b.yaml` (from `template.yaml`) because we already loaded that file.

Reply:
The bug was found and this was resolved so that all tests will pass.